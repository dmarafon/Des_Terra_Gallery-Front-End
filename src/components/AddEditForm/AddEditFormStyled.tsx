import styled from "styled-components";

const AddEditFormStyled = styled.div`
  display: flex;
  text-decoration: none;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
  flex-direction: row;
  @media (max-width: 920px) {
    flex-direction: column;
  }

  .addedit__text--container {
    margin-bottom: 30px;
    display: flex;
    flex-direction: column;
    align-content: flex-start;
    flex-wrap: wrap;
    margin-left: 90px;
    margin-right: 70px;
  }

  .addedit__form--container-labels {
    display: flex;
    flex-direction: row;
    @media (max-width: 920px) {
      flex-direction: column;
    }
  }
  .addedit__form--container {
    display: flex;
  }

  .addedit__container {
    display: flex;
    text-align: center;
    font-family: "Karla", sans-serif;
    text-decoration: none;
    display: flex;
    justify-content: center;
    align-items: center;
    flex-wrap: wrap;
    flex-direction: column;
  }

  .addedit__text--intro {
    color: #000000;
    font-size: 30%;
    text-align: center;
    margin-left: 0.006%;
    margin-bottom: 0;
    margin-top: 20px;
    font-weight: 800;
  }
  .addedit__text--navigation {
    text-decoration: none;
    color: black;
    cursor: pointer;
  }

  .addedit__text--login {
    color: #000000;
    font-size: 30%;
    font-weight: 400;
    text-align: center;
    margin-left: 0.006%;
    margin-top: 60px;
    cursor: pointer;
    font-weight: 700;
  }

  .addedit__text--colored {
    color: #fa803b;
  }

  .addedit__form {
    flex-direction: row;
    flex-wrap: nowrap;
    font-size: 500%;
    font-weight: 700;
    text-align: center;
    display: flex;
    padding: 30px;
    height: auto;
    border-radius: 10px;
    justify-content: center;
    align-items: center;
    flex-wrap: wrap;
    justify-content: space-evenly;
  }

  .addedit__form--first-labels {
    display: flex;
    flex-direction: column;
    margin-right: 40px;
    @media (max-width: 920px) {
      margin-right: 0px;
    }
    flex-basis: 100%;
  }

  .addedit__form--second-labels {
    display: flex;
    flex-direction: column;
    margin-left: 40px;
    @media (max-width: 920px) {
      margin-left: 0px;
    }
    flex-basis: 100%;
  }

  .addedit__label {
    text-align: left;
    font-size: 15px;
    color: #000000;
    cursor: pointer;
  }

  .addedit__label--checkbox {
    text-align: left;
    font-size: 18px;
    color: #000000;
    cursor: pointer;
    margin-left: 1px;
  }

  .addedit__input {
    font-size: 2.2vh;
    font-family: "Karla", sans-serif;
    margin: 15px 0;
    padding: 15px 10px;
    width: 100%;
    outline: none;
    border: 1px solid #000000;
    border-radius: 5px;
    display: inline-block;
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
    -webkit-transition: 0.2s ease all;
    -moz-transition: 0.2s ease all;
    -ms-transition: 0.2s ease all;
    -o-transition: 0.2s ease all;
    transition: 0.2s ease all;
  }

  .addedit__input--checkbox_click {
    width: 2vh;
    height: 2vh;
  }

  .addedit__input[type="text"]:focus,
  .addedit__input[type="password"]:focus {
    border-color: #000000;
  }

  .addedit__input--checkbox {
    display: flex;
  }

  .addedit__button--container {
    display: flex;
    flex-wrap: wrap;
    flex-direction: column;
    align-content: flex-end;
    @media (max-width: 920px) {
      align-content: center;
    }
    margin-top: 40px;
  }

  .addedit__button {
    cursor: pointer;
    justify-content: center;
    height: 50px;
    width: 150px;
    background-color: #fac43b;
    color: #000000;
    border-radius: 5px;
    border-style: none;
    font-family: "Karla", sans-serif;
    font-style: normal;
    font-weight: 700;
    font-size: 17px;
    line-height: 28px;
    opacity: 100%;
    max-width: 100%;
    padding: 11px 15px 10px;
    box-shadow: 5px 5px 10px -3px rgba(0, 0, 0, 0.199),
      -5px -5px 15px 3px rgba(255, 255, 255, 0.199);
    :hover {
      opacity: 100%;
    }
  }

  .addedit__button:hover {
    opacity: 0.8;
  }

  .addedit__button:active {
    opacity: 0.4;
  }
`;

export default AddEditFormStyled;
