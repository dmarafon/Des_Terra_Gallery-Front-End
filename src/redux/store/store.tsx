import { configureStore } from "@reduxjs/toolkit";
import userReducer from "../features/userSlice";
import uiReducer from "../features/uiSlice";

const store = configureStore({
  reducer: {
    user: userReducer,
    ui: uiReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;

export default store;
