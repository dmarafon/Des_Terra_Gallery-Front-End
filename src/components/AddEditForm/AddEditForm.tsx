import { SyntheticEvent, useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import {
  apiResponseActionCreator,
  cleanApiResponseActionCreator,
} from "../../redux/features/uiSlice";
import {
  createArtworkThunk,
  editArtworkThunk,
} from "../../redux/thunks/artworkThunks";
import { loadSingleArtworkThunk } from "../../redux/thunks/singleArtworkThunk";
import { useAppDispatch, useAppSelector } from "../hooks";
import LoadingModal from "../LoadingModal/LoadingModal";
import ModalText from "../ModalText/ModalText";
import AddEditFormStyled from "./AddEditFormStyled";

const AddEditForm = (): JSX.Element => {
  const dispatch = useAppDispatch();

  const {
    title,
    medium,
    height,
    width,
    style,
    purchaseprice,
    monthlyrateprice,
    description,
  } = useAppSelector((state) => state.singleArtwork);

  const { artworkId } = useParams();

  useEffect(() => {
    if (artworkId) {
      dispatch(loadSingleArtworkThunk(artworkId));

      setFormData({
        title: title || "",
        medium: medium || "",
        height: height || "",
        width: width || "",
        style: style || "",
        purchaseprice: purchaseprice || "",
        monthlyrateprice: monthlyrateprice || "",
        description: description || "",
        artimages: "",
      });
    }
  }, [
    artworkId,
    description,
    dispatch,
    height,
    medium,
    monthlyrateprice,
    purchaseprice,
    style,
    title,
    width,
  ]);

  const formInitialState = {
    title: "",
    medium: "",
    height: "",
    width: "",
    style: "",
    purchaseprice: "",
    monthlyrateprice: "",
    description: "",
    artimages: "",
  };

  const [formData, setFormData] = useState(formInitialState);
  const apiMessage = useAppSelector((state) => state.ui.apiResponse);
  const loading = useAppSelector((state) => state.ui.loading);
  const feedback = useAppSelector((state) => state.ui.feedback);
  const singleArtwork = useAppSelector((state) => state.singleArtwork);

  const changeData = (event: SyntheticEvent) => {
    setFormData({
      ...formData,
      [(event.target as HTMLInputElement).id]:
        (event.target as HTMLInputElement).type === "file"
          ? (event.target as HTMLInputElement).files?.[0]
          : (event.target as HTMLInputElement).value,
    });
  };

  const resetForm = () => {
    setFormData(formInitialState);
  };

  const submitCreate = (event: SyntheticEvent) => {
    event.preventDefault();

    switch (true) {
      case formData.title === "":
        dispatch(apiResponseActionCreator("Blank 'Title'"));
        break;
      case formData.medium === "":
        dispatch(apiResponseActionCreator("Blank 'Medium'"));
        break;
      case formData.height === "":
        dispatch(apiResponseActionCreator("Blank 'Height'"));
        break;
      case formData.width === "":
        dispatch(apiResponseActionCreator("Blank 'Width'"));
        break;
      case formData.style === "":
        dispatch(apiResponseActionCreator("Blank 'Style'"));
        break;
      case formData.purchaseprice === "":
        dispatch(apiResponseActionCreator("Blank 'Purchase Price'"));
        break;
      case formData.monthlyrateprice === "":
        dispatch(apiResponseActionCreator("Blank 'Monthly Rate Price'"));
        break;

      case formData.description === "":
        dispatch(apiResponseActionCreator("Blank 'Description'"));
        break;

      default:
        const newArtwork = new FormData();
        newArtwork.append("title", formData.title.toLowerCase());
        newArtwork.append("medium", formData.medium.toLowerCase());
        newArtwork.append("height", formData.height);
        newArtwork.append("width", formData.width);
        newArtwork.append("style", formData.style.toLowerCase());
        newArtwork.append(
          "purchaseprice",
          formData.purchaseprice.toLowerCase()
        );
        newArtwork.append(
          "monthlyrateprice",
          formData.monthlyrateprice.toLowerCase()
        );
        newArtwork.append("description", formData.description);
        newArtwork.append("artimages", formData.artimages);

        if (artworkId) {
          dispatch(editArtworkThunk(newArtwork, artworkId));
        } else {
          dispatch(createArtworkThunk(newArtwork));
        }
        setFormData(formInitialState);
        resetForm();
    }
  };

  const artTitle = singleArtwork.title;

  const submitClosingModalResponse = () => {
    dispatch(cleanApiResponseActionCreator());
  };

  return (
    <>
      {apiMessage === "Unknown Error" && (
        <ModalText
          handleClose={submitClosingModalResponse}
          isOpen={false}
          customFunction={""}
        >
          Oops... We're sorry, something went wrong with our servers, try again
          later
          <p className="login__modal--break_text"></p>
        </ModalText>
      )}
      {apiMessage.startsWith("Blank") && (
        <ModalText
          handleClose={submitClosingModalResponse}
          isOpen={feedback}
          customFunction={""}
        >
          Please, you left the {apiMessage.substring(6)} mandatory field in
          blank. All fields are mandatory.
        </ModalText>
      )}
      {apiMessage === "new" && (
        <ModalText
          handleClose={submitClosingModalResponse}
          isOpen={feedback}
          customFunction={""}
        >
          Your Artwork was Created Succesfully! You will be redirected to your
          Art page
        </ModalText>
      )}
      {apiMessage === "upd" && (
        <ModalText
          handleClose={submitClosingModalResponse}
          isOpen={feedback}
          customFunction={""}
        >
          Your Artwork was Updated Succesfully! You will be redirected to your
          Art page
        </ModalText>
      )}
      {apiMessage === "Bad Request" && (
        <ModalText
          handleClose={submitClosingModalResponse}
          isOpen={feedback}
          customFunction={""}
        >
          We're sorry, something went wrong
          <p className="login__modal--break_text">
            {" "}
            Please, try again to Create your Art
          </p>
        </ModalText>
      )}
      {loading ? (
        <LoadingModal />
      ) : (
        <AddEditFormStyled>
          <div>
            <div className="addedit__form">
              <form onSubmit={submitCreate} noValidate autoComplete="off">
                {artworkId ? (
                  <div className="addedit__text--container">
                    <p className="addedit__text--intro">EDIT YOUR WORK</p>
                    <p className="addedit__text--intro">
                      {" "}
                      <span className="addedit__text--colored-edit">
                        {`"${artTitle}"`}
                      </span>{" "}
                    </p>
                  </div>
                ) : (
                  <div className="addedit__text--container">
                    <p className="addedit__text--intro">ADD YOUR WORK</p>
                    <p className="addedit__text--intro">
                      TO START SELLING AND{" "}
                    </p>
                    <p className="addedit__text--intro">
                      RENTING
                      <span className="addedit__text--colored">
                        {" "}
                        RIGHT NOW!
                      </span>
                    </p>
                  </div>
                )}
                <div className="addedit__form--container-labels">
                  <div className="addedit__form--first-labels">
                    <label className="addedit__label" htmlFor="title">
                      TITLE
                    </label>
                    <input
                      className="addedit__input"
                      id="title"
                      placeholder="TITLE"
                      value={formData.title}
                      onChange={changeData}
                      required
                      autoComplete="off"
                      maxLength={24}
                    />
                    <label className="addedit__label" htmlFor="medium">
                      MEDIUM
                    </label>
                    <input
                      className="addedit__input"
                      id="medium"
                      placeholder="MEDIUM"
                      value={formData.medium}
                      onChange={changeData}
                      required
                      autoComplete="off"
                      maxLength={33}
                    />
                    <label className="addedit__label" htmlFor="height">
                      HEIGHT (DIMENSIONS)
                    </label>
                    <input
                      className="addedit__input"
                      id="height"
                      type="number"
                      min="1"
                      placeholder="HEIGHT"
                      value={formData.height}
                      onChange={changeData}
                      required
                      autoComplete="off"
                      maxLength={4}
                    />
                    <label className="addedit__label" htmlFor="width">
                      WIDTH (DIMENSIONS)
                    </label>
                    <input
                      className="addedit__input"
                      id="width"
                      placeholder="WIDTH"
                      value={formData.width}
                      onChange={changeData}
                      required
                      autoComplete="off"
                      type="number"
                      min="1"
                      maxLength={4}
                    />
                    <label className="addedit__label" htmlFor="style">
                      STYLE
                    </label>
                    <select
                      id="style"
                      className="addedit__select"
                      name="style"
                      required
                      value={formData.style}
                      onChange={changeData}
                    >
                      <option disabled value="">
                        Select a Style
                      </option>
                      <option value="abstract">Abstract</option>
                      <option value="architecture">Architecture</option>
                      <option value="expressionist">Expressionist</option>
                      <option value="figurative">Figurative</option>
                      <option value="geometric">Geometric</option>
                      <option value="minimal">Minimal</option>
                      <option value="nature">Nature</option>
                      <option value="people">People</option>
                      <option value="psychedelic">Psychedelic</option>
                      <option value="urban">Urban</option>
                    </select>
                  </div>
                  <div className="addedit__form--second-labels">
                    <label className="addedit__label" htmlFor="purchaseprice">
                      PURCHASE PRICE
                    </label>
                    <input
                      className="addedit__input"
                      id="purchaseprice"
                      placeholder="€"
                      value={formData.purchaseprice}
                      required
                      onChange={changeData}
                      autoComplete="off"
                      type="number"
                      min="1"
                      maxLength={10}
                    />
                    <label
                      className="addedit__label"
                      htmlFor="monthlyrateprice"
                    >
                      MONTHLY RATE PRICE
                    </label>
                    <input
                      className="addedit__input"
                      id="monthlyrateprice"
                      placeholder="€"
                      value={formData.monthlyrateprice}
                      onChange={changeData}
                      autoComplete="off"
                      required
                      type="number"
                      min="1"
                      maxLength={10}
                    />
                    <label className="addedit__label" htmlFor="description">
                      DESCRIPTION (MAX 240 CHAR)
                    </label>
                    <textarea
                      maxLength={240}
                      className="addedit__textarea"
                      id="description"
                      placeholder="Describe your Artwork here"
                      value={formData.description}
                      onChange={changeData}
                      autoComplete="off"
                      required
                    />
                    <label className="addedit__label" htmlFor="artimages">
                      ARTWORK PICTURE
                    </label>
                    <input
                      className="addedit__input"
                      id="artimages"
                      type="file"
                      onChange={changeData}
                      autoComplete="off"
                      accept="image/*"
                      required
                    />
                  </div>
                </div>
                <div className="addedit__button--container">
                  {artworkId ? (
                    <button
                      className="addedit__button"
                      type="submit"
                      disabled={false}
                      value="Send"
                    >
                      EDIT YOUR ART
                    </button>
                  ) : (
                    <button
                      className="addedit__button"
                      type="submit"
                      disabled={false}
                      value="Send"
                    >
                      ADD YOUR ART
                    </button>
                  )}
                </div>
              </form>
            </div>
          </div>
        </AddEditFormStyled>
      )}
    </>
  );
};

export default AddEditForm;
