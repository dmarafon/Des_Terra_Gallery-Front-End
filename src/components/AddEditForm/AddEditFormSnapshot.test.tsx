import { Provider } from "react-redux";
import TestRenderer from "react-test-renderer";
import store from "../../redux/store/store";
import AddEditForm from "./AddEditForm";

describe("Given an AddEdit form component", () => {
  describe("When it's rendered", () => {
    test("Then it should always match this snapshot", () => {
      const renderedAddEditForm = TestRenderer.create(
        <Provider store={store}>
          <AddEditForm />
        </Provider>
      );

      expect(renderedAddEditForm).toMatchSnapshot();
    });
  });
});
