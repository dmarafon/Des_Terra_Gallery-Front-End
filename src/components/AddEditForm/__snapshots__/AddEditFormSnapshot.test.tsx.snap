// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given an AddEdit form component When it's rendered Then it should always match this snapshot 1`] = `
<div
  className="sc-dkzDqf ebvruw"
>
  <div>
    <div
      className="addedit__form"
    >
      <form
        autoComplete="off"
        noValidate={true}
        onSubmit={[Function]}
      >
        <div
          className="addedit__text--container"
        >
          <p
            className="addedit__text--intro"
          >
            ADD YOUR WORK
          </p>
          <p
            className="addedit__text--intro"
          >
            TO START SELLING AND
             
          </p>
          <p
            className="addedit__text--intro"
          >
            RENTING
            <span
              className="addedit__text--colored"
            >
               
              RIGHT NOW!
            </span>
          </p>
        </div>
        <div
          className="addedit__form--container-labels"
        >
          <div
            className="addedit__form--first-labels"
          >
            <label
              className="addedit__label"
              htmlFor="title"
            >
              TITLE
            </label>
            <input
              autoComplete="off"
              className="addedit__input"
              id="title"
              maxLength={24}
              onChange={[Function]}
              placeholder="TITLE"
              required={true}
              value=""
            />
            <label
              className="addedit__label"
              htmlFor="medium"
            >
              MEDIUM
            </label>
            <input
              autoComplete="off"
              className="addedit__input"
              id="medium"
              maxLength={33}
              onChange={[Function]}
              placeholder="MEDIUM"
              required={true}
              value=""
            />
            <label
              className="addedit__label"
              htmlFor="height"
            >
              HEIGHT (DIMENSIONS)
            </label>
            <input
              autoComplete="off"
              className="addedit__input"
              id="height"
              maxLength={4}
              min="1"
              onChange={[Function]}
              placeholder="HEIGHT"
              required={true}
              type="number"
              value=""
            />
            <label
              className="addedit__label"
              htmlFor="width"
            >
              WIDTH (DIMENSIONS)
            </label>
            <input
              autoComplete="off"
              className="addedit__input"
              id="width"
              maxLength={4}
              min="1"
              onChange={[Function]}
              placeholder="WIDTH"
              required={true}
              type="number"
              value=""
            />
            <label
              className="addedit__label"
              htmlFor="style"
            >
              STYLE
            </label>
            <select
              className="addedit__select"
              id="style"
              name="style"
              onChange={[Function]}
              required={true}
              value=""
            >
              <option
                disabled={true}
                value=""
              >
                Select a Style
              </option>
              <option
                value="abstract"
              >
                Abstract
              </option>
              <option
                value="architecture"
              >
                Architecture
              </option>
              <option
                value="expressionist"
              >
                Expressionist
              </option>
              <option
                value="figurative"
              >
                Figurative
              </option>
              <option
                value="geometric"
              >
                Geometric
              </option>
              <option
                value="minimal"
              >
                Minimal
              </option>
              <option
                value="nature"
              >
                Nature
              </option>
              <option
                value="people"
              >
                People
              </option>
              <option
                value="psychedelic"
              >
                Psychedelic
              </option>
              <option
                value="urban"
              >
                Urban
              </option>
            </select>
          </div>
          <div
            className="addedit__form--second-labels"
          >
            <label
              className="addedit__label"
              htmlFor="purchaseprice"
            >
              PURCHASE PRICE
            </label>
            <input
              autoComplete="off"
              className="addedit__input"
              id="purchaseprice"
              maxLength={10}
              min="1"
              onChange={[Function]}
              placeholder="€"
              required={true}
              type="number"
              value=""
            />
            <label
              className="addedit__label"
              htmlFor="monthlyrateprice"
            >
              MONTHLY RATE PRICE
            </label>
            <input
              autoComplete="off"
              className="addedit__input"
              id="monthlyrateprice"
              maxLength={10}
              min="1"
              onChange={[Function]}
              placeholder="€"
              required={true}
              type="number"
              value=""
            />
            <label
              className="addedit__label"
              htmlFor="description"
            >
              DESCRIPTION (MAX 240 CHAR)
            </label>
            <textarea
              autoComplete="off"
              className="addedit__textarea"
              id="description"
              maxLength={240}
              onChange={[Function]}
              placeholder="Describe your Artwork here"
              required={true}
              value=""
            />
            <label
              className="addedit__label"
              htmlFor="artimages"
            >
              ARTWORK PICTURE
            </label>
            <input
              accept="image/*"
              autoComplete="off"
              className="addedit__input"
              id="artimages"
              onChange={[Function]}
              required={true}
              type="file"
            />
          </div>
        </div>
        <div
          className="addedit__button--container"
        >
          <button
            className="addedit__button"
            disabled={false}
            type="submit"
            value="Send"
          >
            ADD YOUR ART
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

